unit/test-rilmodem-cf.c

```diff
@@ -0,0 +1,513 @@
+	0x33, 0x00, 0x34, 0x00, 0x35, 0x00, 0x36, 0x00, 0x37, 0x00, 0x38, 0x00,
+	0x39, 0x00, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x3c, 0x00, 0x00, 0x00
+};
+
+static const struct cf_data testdata_cf_set_valid_4 = {
+	.call_type = CF_CALL_ERASURE,
+	.rtd = {
+		.req_data = req_cf_set_parcel_4,
+		.req_size = sizeof(req_cf_set_parcel_4),
+		.rsp_data = NULL,
+		.rsp_size = 0,
+		.rsp_error = RIL_E_SUCCESS,
+	},
+	.type = 0,
+	.cls = 0,
+	.number = { "1234567890", OFONO_NUMBER_TYPE_UNKNOWN },
+	.time = 60,
+	.error_type = OFONO_ERROR_TYPE_NO_ERROR,
+	.total = 0,
+	.list = NULL,
+};
+
+/* Test SET_CALL_FORWARD, error on activate unconditional forwarding */
+static const guchar req_cf_set_parcel_inv_1[] = {
+	0x00, 0x00, 0x00, 0x38, 0x22, 0x00, 0x00, 0x00, 0x49, 0x01, 0x00, 0x00,
+	0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x81, 0x00, 0x00, 0x00, 0x0a, 0x00, 0x00, 0x00, 0x31, 0x00, 0x32, 0x00,
+	0x33, 0x00, 0x34, 0x00, 0x35, 0x00, 0x36, 0x00, 0x37, 0x00, 0x38, 0x00,
+	0x39, 0x00, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x3c, 0x00, 0x00, 0x00
+};
+
+static const struct cf_data testdata_cf_set_invalid_1 = {
+	.call_type = CF_CALL_ACTIVATION,
+	.rtd = {
+		.req_data = req_cf_set_parcel_inv_1,
+		.req_size = sizeof(req_cf_set_parcel_inv_1),
+		.rsp_data = NULL,
+		.rsp_size = 0,
+		.rsp_error = RIL_E_GENERIC_FAILURE,
+	},
+	.type = 0,
+	.cls = 0,
+	.number = { "1234567890", OFONO_NUMBER_TYPE_UNKNOWN },
+	.time = 60,
+	.error_type = OFONO_ERROR_TYPE_FAILURE,
+	.total = 0,
+	.list = NULL,
+};
+
+/* Test QUERY_CALL_FORWARD_STATUS. Real transaction from turbo. */
+static const guchar req_cf_query_parcel_1[] = {
+	0x00, 0x00, 0x00, 0x38, 0x21, 0x00, 0x00, 0x00, 0xa5, 0x00, 0x00, 0x00,
+	0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x81, 0x00, 0x00, 0x00, 0x0a, 0x00, 0x00, 0x00, 0x31, 0x00, 0x32, 0x00,
+	0x33, 0x00, 0x34, 0x00, 0x35, 0x00, 0x36, 0x00, 0x37, 0x00, 0x38, 0x00,
+	0x39, 0x00, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x3c, 0x00, 0x00, 0x00
+};
+
+/* reply data for QUERY_CALL_FORWARD (disabled) */
+static const guchar rsp_cf_query_data_1[] = {
+	0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+};
+
+struct ofono_call_forwarding_condition list_cf_query_valid_1 = {
+	.status = 0,
+	.cls = 7,
+	.phone_number = { "", 0 },
+	.time = 0
+};
+
+static const struct cf_data testdata_cf_query_valid_1 = {
+	.call_type = CF_CALL_QUERY,
+	.rtd = {
+		.req_data = req_cf_query_parcel_1,
+		.req_size = sizeof(req_cf_query_parcel_1),
+		.rsp_data = rsp_cf_query_data_1,
+		.rsp_size = sizeof(rsp_cf_query_data_1),
+		.rsp_error = RIL_E_SUCCESS,
+	},
+	.type = 0,
+	.cls = 0,
+	.error_type = OFONO_ERROR_TYPE_NO_ERROR,
+	.total = 1,
+	.list = &list_cf_query_valid_1,
+};
+
+/* Test QUERY_CALL_FORWARD_STATUS, return error */
```

Comment:
Change "return error" to "GENERIC_FAILURE returned in RIL reply" ( or add this just ahead of the testdata ).  See test-rilodem-cs.c line 280 for an example.